GeneratorInterface/Herwig7Interface/src/Herwig7Interface.cc

```diff
@@ -0,0 +1,347 @@
+#include <ThePEG/EventRecord/Particle.h> 
+#include <ThePEG/EventRecord/Collision.h>
+#include <ThePEG/EventRecord/TmpTransform.h>
+#include <ThePEG/Config/ThePEG.h>
+#include <ThePEG/PDF/PartonExtractor.h>
+#include <ThePEG/PDF/PDFBase.h>
+#include <ThePEG/Utilities/UtilityBase.h>
+#include <ThePEG/Vectors/HepMCConverter.h>
+
+
+#include "FWCore/MessageLogger/interface/MessageLogger.h"
+
+#include "GeneratorInterface/Core/interface/ParameterCollector.h"
+
+#include "GeneratorInterface/Herwig7Interface/interface/Proxy.h"
+#include "GeneratorInterface/Herwig7Interface/interface/RandomEngineGlue.h"
+#include "GeneratorInterface/Herwig7Interface/interface/Herwig7Interface.h"
+
+#include "CLHEP/Random/RandomEngine.h"
+
+using namespace std;
+using namespace gen;
+
+Herwig7Interface::Herwig7Interface(const edm::ParameterSet &pset) :
+	randomEngineGlueProxy_(ThePEG::RandomEngineGlue::Proxy::create()),
+	dataLocation_(ParameterCollector::resolve(pset.getParameter<string>("dataLocation"))),
+	generator_(pset.getParameter<string>("generatorModule")),
+	run_(pset.getParameter<string>("run")),
+	dumpConfig_(pset.getUntrackedParameter<string>("dumpConfig", "HerwigConfig.in")),
+	skipEvents_(pset.getUntrackedParameter<unsigned int>("skipEvents", 0))
+{
+	// Write events in hepmc ascii format for debugging purposes
+	string dumpEvents = pset.getUntrackedParameter<string>("dumpEvents", "");
+	if (!dumpEvents.empty()) {
+		iobc_.reset(new HepMC::IO_GenEvent(dumpEvents.c_str(), ios::out));
+		edm::LogInfo("ThePEGSource") << "Event logging switched on (=> " << dumpEvents << ")";
+	}
+	// Clear dumpConfig target
+	if (!dumpConfig_.empty())
+		ofstream cfgDump(dumpConfig_.c_str(), ios_base::trunc);
+}
+
+Herwig7Interface::~Herwig7Interface () noexcept
+{
+	if (eg_)
+		eg_->finalize();
+	edm::LogInfo("Herwig7Interface") << "Event generator finalized";
+}
+
+void Herwig7Interface::setPEGRandomEngine(CLHEP::HepRandomEngine* v) {
+    
+        randomEngineGlueProxy_->setRandomEngine(v);
+        randomEngine = v;
+        ThePEG::RandomEngineGlue *rnd = randomEngineGlueProxy_->getInstance();
+        if(rnd) {
+            rnd->setRandomEngine(v);
+        }
+}
+
+
+void Herwig7Interface::initRepository(const edm::ParameterSet &pset)
+{
+	std::string runModeTemp = pset.getUntrackedParameter<string>("runModeList","read,run");
+
+	// To Lower
+	std::transform(runModeTemp.begin(), runModeTemp.end(), runModeTemp.begin(), ::tolower);
+
+
+
+	while(!runModeTemp.empty())
+	{
+		// Split first part of List
+		std::string choice;
+		size_t pos = runModeTemp.find(",");
+        if (std::string::npos == pos)
+            choice=runModeTemp;
+        else
+		    choice = runModeTemp.substr(0, pos);
+        
+		if (pos == std::string::npos)
+			runModeTemp.erase();
+        else
+            runModeTemp.erase(0, pos+1);
+
+		// construct HerwigUIProvider object and return it as global object
+		HwUI_ = new Herwig::HerwigUIProvider(pset, dumpConfig_, Herwig::RunMode::READ);
```

Comment:
This looks like a memory leak.

Reply:
I added a delete option for HwUI_ object in case it already exists